### PR TITLE
WIP: Fix CLI use for source-only rosdistros.

### PR DESCRIPTION
**Pull request by mikepurvis**
Since we finished the process of transitioning away from using Bloom, we no longer have release stanzas in our rosdistro repo— it's exclusively tagged source repos. I noticed today that under these circumstances, `rosinstall_generator` was broken at the command line, which we hadn't noticed previously because our build tooling uses it by importing and calling the underlying functions directly (which among other things bypasses a lot of the wet/dry compatibility stuff).

Anyway, this fix allows the following invocations to work on a source-only rosdistro where an appropriate source cache is available. Example specifying repos with dependencies:

```
$ rosinstall_generator --rosdistro indigo --upstream-development --deps --repos ros_comm
- git:
    local-name: catkin
    uri: https://github.com/ros/catkin.git
    version: 0.7.8
[...]
- git:
    local-name: ros_comm
    uri: https://github.com/ros/ros_comm.git
    version: e69b098bd4d80e7c8feda1eb1a9aaf67cdac352e
- git:
    local-name: ros_comm_msgs
    uri: https://github.com/ros/ros_comm_msgs.git
    version: 1.11.2
- git:
    local-name: roscpp_core
    uri: https://github.com/ros/roscpp_core.git
    version: 0.6.7
- git:
    local-name: roslisp
    uri: https://github.com/ros/roslisp.git
    version: 1.9.21
- git:
    local-name: rospack
    uri: https://github.com/ros/rospack.git
    version: 2.2.8
- git:
    local-name: std_msgs
    uri: https://github.com/ros/std_msgs.git
    version: 0.5.11
```

Example specifying exact package with deps, getting tarballs:

```
$ rosinstall_generator --rosdistro indigo --upstream-development --deps --tar rospack
- tar:
    local-name: catkin
    uri: https://github.com/ros/catkin/archive/0.7.8.tar.gz
    version: catkin-0.7.8
- tar:
    local-name: cmake_modules
    uri: https://github.com/ros/cmake_modules/archive/0.4.1.tar.gz
    version: cmake_modules-0.4.1
- tar:
    local-name: rospack
    uri: https://github.com/ros/rospack/archive/2.2.8.tar.gz
    version: rospack-2.2.8
```

I've also validated that this change doesn't break the `--upstream-development` flag when used with a rosdistro cache that does not include the source package xml cache (such as the one at `http://repositories.ros.org/`).

### CODE REVIEW

**Comment by sloretz** [comment]
Hi @mikepurvis, Thanks for the PR! I won't have time to look at this for a little while.

Would you be willing to add a test for this?

File: src/rosinstall_generator/generator.py [hunk 1]
```diff
@@ -87,20 +87,27 @@ def _classify_repo_names(distro_name, repo_names):
...
     wet_distro = get_wet_distro(distro_name)
     for repo_name in repo_names:
-        release_repo = wet_distro.repositories[repo_name].release_repository
-        if release_repo:
-            package_names.update(release_repo.package_names)
+        if source:
+            source_package_xmls = wet_distro.get_source_repo_package_xmls(repo_name)
+        if source and source_package_xmls:
+            source_package_names = source_package_xmls.keys()
+            source_package_names.remove("_ref")
```
Comment (dirk-thomas) [inline src/rosinstall_generator/generator.py]: Can you please add a comment above this line why it is necessary. It took quite a bit to figure it out when reading the patch.

**Comment by mikepurvis** [comment]
@sloretz Sure, but we should talk about what kind of testing strategy is desired or appropriate for this repo. Currently, there's a _lot_ of code in here which handles corner cases pertaining to rosbuild packages (which I have no interest in attempting to support or test support for), and there's also a lot of dependence on network access, which would be pretty tricky to mock out in a way that retained the legitimacy of the test itself.

Basically, would you be interested in a suite which was based on having some separate mini test rosdistros (one source-only, one source/release, one release-only) consisting of tags/hashes to repos, where the suite would build the relevant caches and then exercise some standard rosinstall_generator operations against those? Such a test would be heavily network/github dependent, but I think that's just a reality of this code— and still a huge step up from the current state affairs, which is no tests at all.

File: src/rosinstall_generator/generator.py (hunk 1 above)
Comment (mikepurvis) [reply src/rosinstall_generator/generator.py]: Added, also filed some related tickets against `rosdistro` which would help this in the future:

https://github.com/ros-infrastructure/rosdistro/issues/115

https://github.com/ros-infrastructure/rosdistro/issues/116

**Comment by mikepurvis** [comment]
It's probably a fairly small fix, and making it now would avoid having to do a compatibility dance to change it down the road. If you guys are okay with merging both PRs and coordinating the next release, then I'll put up one for rosdistro, and then this one will be updated to remove the hack and bump up the version dependency.

**Comment by dirk-thomas** [comment]
> If you guys are okay with merging both PRs and coordinating the next release, then I'll put up one for rosdistro, and then this one will be updated to remove the hack and bump up the version dependency.

Absolutely. Thanks.

**Comment by mikepurvis** [comment]
Okay, this is updated now, but obviously this shouldn't merge until https://github.com/ros-infrastructure/rosdistro/pull/117 is merged and released.

File: setup.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 setup(
     name='rosinstall_generator',
     version=__version__,
-    install_requires=['argparse', 'catkin_pkg >= 0.1.28', 'rosdistro >= 0.5.0', 'rospkg', 'PyYAML', 'setuptools'],
+    install_requires=['argparse', 'catkin_pkg >= 0.1.28', 'rosdistro >= 0.6.7', 'rospkg', 'PyYAML', 'setuptools'],
```
Comment (dirk-thomas) [inline setup.py]: The version also needs to be updated in these two lines in the `stdeb.cfg` file: https://github.com/ros-infrastructure/rosinstall_generator/blob/ce4e6ec834d1c7238ebfaf7243d355bde3b5676e/stdeb.cfg#L2-L3

Comment (mikepurvis) [reply setup.py]: Fixed.

Comment (mikepurvis) [reply setup.py]: Conflict introduced by #45 is also resolved now.

File: setup.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 import sys
 from setuptools import setup, find_packages
 
-install_requires=['catkin_pkg >= 0.1.28', 'rosdistro >= 0.5.0', 'rospkg', 'PyYAML', 'setuptools']
+install_requires=['catkin_pkg >= 0.1.28', 'rosdistro >= 0.6.7', 'rospkg', 'PyYAML', 'setuptools']
```
Comment (dirk-thomas) [inline setup.py]: Do you want to bump the minimum version to 0.6.8 (same below)?

Comment (mikepurvis) [reply setup.py]: Done.

**Comment by dirk-thomas** [comment]
Thanks. :ship:
